src/passes/OptimizeInstructions.cpp

```diff
@@ -1372,39 +1372,63 @@ struct OptimizeInstructions
     }
   }
 
-  void visitRefEq(RefEq* curr) {
-    // Equality does not depend on the type, so casts may be removable.
-    skipCast(curr->left, Type::eqref);
-    skipCast(curr->right, Type::eqref);
-
-    // Identical references compare equal.
-    // (Technically we do not need to check if the inputs are also foldable into
-    // a single one, but we do not have utility code to handle non-foldable
-    // cases yet; the foldable case we do handle is the common one of the first
-    // child being a tee and the second a get of that tee. TODO)
-    if (areConsecutiveInputsEqualAndFoldable(curr->left, curr->right)) {
-      auto* result =
-        Builder(*getModule()).makeConst(Literal::makeOne(Type::i32));
-      replaceCurrent(getDroppedChildrenAndAppend(
-        curr, *getModule(), getPassOptions(), result));
-      return;
-    }
-
-    // Canonicalize to the pattern of a null on the right-hand side, if there is
-    // one. This makes pattern matching simpler.
-    if (curr->left->is<RefNull>()) {
-      std::swap(curr->left, curr->right);
-    }
-
-    // RefEq of a value to Null can be replaced with RefIsNull.
-    if (curr->right->is<RefNull>()) {
-      replaceCurrent(Builder(*getModule()).makeRefIs(RefIsNull, curr->left));
-    }
-  }
+  // Note on removing casts (which the following utilities, skipNonNullCast and
+  // skipCast do): removing a cast is potentially dangerous, as it removes
+  // information from the IR. For example:
+  //
+  //  (ref.is_func
+  //    (ref.as_func
+  //      (local.get $anyref)))
+  //
+  // The local has no useful type info here (it is anyref). The cast forces it
+  // to be a function, so we know that if we do not trap then the ref.is will
+  // definitely be 1. But if we removed the ref.as first (which we can do in
+  // traps-never-happen mode) then we'd not have the type info we need to
+  // optimize that way.
+  //
+  // To avoid such risks we should keep in mind the following:
+  //
+  //  * Before removing a cast we should use its type information in the best
+  //    way we can. Only after doing so should a cast be removed. In the exmaple
+  //    above, that means first seeing that the ref.is must return 1, and only
+  //    then possibly removing the ref.as.
+  //  * Do not remove a cast if removing it might remove useful information for
+  //    others. For example,
+  //
+  //      (ref.cast $A
+  //        (ref.as_non_null ..))
+  //
+  //    If we remove the inner cast then the outer cast becomes nullable. That
```

Comment:
Doesn't `ref.cast $A`'s type depend solely on `$A` type? Does whether we have `ref.as_non_null` inside determine whether the outer cast `ref.cast` becomes nullable or not?

Reply:
Yes, the heap type of `(ref.cast $A ..)` is `$A`, correct, while it can be nullable or non-nullable (which it copies from the child - if the child is nullable, it is too).